### PR TITLE
[FLOC-4423] A utility to assist in generation tracking.

### PR DESCRIPTION
**Pull request by sarum90**
This PR creates a utility that will be used in the next PR to track the current generation of the configuration and state, and be able to upgrade from older versions of the config and state to the latest version.

### CODE REVIEW

File: flocker/control/_generations.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        ``_GenerationRecord`` to the queue (and drop one if the queue is full).
+        """
+        if latest == self._latest_object:
+            return 
+        latest_hash = make_generation_hash(latest)
+
+        if (self._latest_object is not None and
+                latest_hash != self._latest_hash):
+            new_diff = create_diff(self._latest_object, latest)
+            if self._latest_hash is not None:
```
Comment (wallrj) [inline flocker/control/_generations.py]: :question: I can't think why this check is necessary? `make_generation_hash` always returns a `GenerationHash`.

Comment (wallrj) [inline flocker/control/_generations.py]: Coverage report agrees.

File: flocker/control/_model.py [hunk 2]
```diff
@@ -1232,10 +1232,50 @@ def get_information_wipe(self):
...
+
+
+class GenerationHash(PClass):
+    """
+    Generation hash uniquely defines a configuration or state. A value of None
+    indicates that there is no configuration or state known.
+
+    Note that this is a tuple of ints rather than ``bytes`` because the flocker
+    ``wire_encode`` function assumes that ``bytes`` can be encoded as JSON
+    strings, which is not true of generation hashes.
```
Comment (wallrj) [inline flocker/control/_model.py]: :question: Did you consider `base64` encoding instead?

File: flocker/control/test/test_diffing.py [hunk 3]
```diff
@@ -54,6 +56,24 @@ def test_deployment_diffing(self, deployment_a, deployment_b):
             Equals(deployment_b)
         )
 
+    @given(
+        st.lists(deployment_strategy(), min_size=3, max_size=10)
+    )
+    def test_deployment_diffing_composable(self, deployments):
+        """
+        Diffs should compose to create an aggregate diff.
```
Comment (wallrj) [inline flocker/control/test/test_diffing.py]: :question: Haha, It took me a while to get my head around this test. Add a little comment outlining the strategy for my benefit.

File: flocker/control/test/test_generations.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    def test_basic_use_works(self):
+        """
+        After inserting a bunch of deployments into a ``GenerationTracker`` in
+        sequence, the ``Diff`` returned from ``get_diff_from_hash_to_latest``
+        can be applied to convert each of the deployments to the latest
+        deployment.
+        """
+        deployments = list(deployment_strategy().example() for _ in xrange(5))
+        deployments[3] = deployments[1]
```
Comment (wallrj) [inline flocker/control/test/test_generations.py]: :question: It's not obvious to me why ^^ ?. Add a comment.

Comment (wallrj) [inline flocker/control/test/test_generations.py]: :question: Yeah, you'll have to explain your strategy  / intent in this test too. I can't work it out :confused:

File: flocker/control/testtools.py [hunk 5]
```diff
@@ -312,3 +353,22 @@ def deployment_strategy(draw, min_number_of_nodes=1):
...
+
+    :param int number_of_deployments: The number of deployments to create.
+
+    :returns: A strategy to create ``number_of_deployments`` ``Deployment`` s.
+    """
+    node_uuid_pool = draw(node_uuid_pool_strategy())
+    return tuple(
+        draw(deployment_strategy(node_uuid_pool=node_uuid_pool))
+        for _ in xrange(number_of_deployments)
+    )
```
Comment (wallrj) [inline flocker/control/testtools.py]: Cool.

File: flocker/control/test/test_generations.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+class GenerationTrackerTests(TestCase):
+
+    def test_basic_use_works(self):
+        """
+        After inserting a bunch of deployments into a ``GenerationTracker`` in
+        sequence, the ``Diff`` returned from ``get_diff_from_hash_to_latest``
+        can be applied to convert each of the deployments to the latest
+        deployment.
+        """
+        deployments = list(deployment_strategy().example() for _ in xrange(5))
```
Comment (wallrj) [inline flocker/control/test/test_generations.py]: :question: Can you use `related_deployments_strategy` here?

**Comment by wallrj** [comment]
Looks great @sarum90 

It's taken me a while to get my head around the code in this branch, especially some of the tests.
Consider adding some further explanatory docstrings and comments and then merge.

File: flocker/control/_generations.py (hunk 1 above)
Comment (sarum90) [inline flocker/control/_generations.py]: Removed, Thanks!

That was left over from an earlier version of this object that was harder to reason about.

File: flocker/control/_model.py (hunk 2 above)
Comment (sarum90) [inline flocker/control/_model.py]: Yea, that'd be equivalent, and would serialize to something smaller as well, I guess then in make_generation_hash I could compute the base64 encoding and put it in this object. I don't think there is a strong argument to go for one over the other.

On 1 hand base64 serializes smaller, on the other hand python doesn't have a good type to verify that a value is a base64 string... (I guess I could write a pyrsistent invariant or something...)

File: flocker/control/test/test_diffing.py (hunk 3 above)
Comment (sarum90) [inline flocker/control/test/test_diffing.py]: Added.

File: flocker/control/test/test_generations.py (hunk 6 above)
Comment (sarum90) [inline flocker/control/test/test_generations.py]: Lol, yea I wrote that strategy specifically to use it in locations like this, and then just forgot about it :p. Using that instead.

File: flocker/control/test/test_generations.py (hunk 4 above)
Comment (sarum90) [inline flocker/control/test/test_generations.py]: Yea... it was kind of convoluted. Added comments.

The code above is a bit more interesting if the queue has repeated deployments (the line labelled `if record.generation_hash == generation_hash:` gets hit multiple times).

Most of the other things are just trying to double check that inserting the same object as the latest a bunch is a no-op, and that you always get the same diff out of the tracker.

File: flocker/control/testtools.py (hunk 5 above)
Comment (sarum90) [inline flocker/control/testtools.py]: Thanks.

Yea I like this idea in general... if I had more time and thought this code was long-term important I would have done more strategies like this... generate a pool of node uuids, generate a pool of dataset ids, generate a pool of applications, create a bunch of deployments from that, etc.

**Comment by sarum90** [comment]
Added comments, waiting for tests.

I _think_ this might all get a bit clearer with my next PR.... which I'll throw together now :).

**Comment by sarum90** [comment]
The only most recent failure was because Jenkins killed the slave. I've seen all of the other tests pass with the code that is running, so I'm going to merge.

Thanks for the review Richard!

**Pull request merged**